### PR TITLE
[ASTextNode] placeholder image shouldn't draw for zero area strings

### LINKED ISSUE #1986: [ASTextNode] placeholder image context fails for zero width size
CGContextSetFillColorWithColor: invalid context 0x0. If you want to see the backtrace, please set CG_CONTEXT_SHOW_BACKTRACE environmental variable

      Jul 26 15:18:02  BroEx[6115] <Error>: CGContextSetFillColorWithColor: invalid context 0x0. Backtrace:
	  <CGContextSetFillColorWithColor+66>
	   <-[UIDeviceWhiteColor setFill]+55>
	    <-[ASTextNode placeholderImage]+357>
	     <-[ASDisplayNode _completeLayoutCalculation]+558>
	      <-[ASDisplayNode measureWithSizeRange:]+794>
	       <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	        <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objec
	         <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChildrenA
	          <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obje
	           <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0_9allo
	            <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	          Jul 26 15:18:02  BroEx[6115] <Error>: CGContextSetFillColorWithColor: invalid context 0x0. Backtrace:
	  <CGContextSetFillColorWithColor+66>
	   <-[UIDeviceWhiteColor setFill]+55>
	    <-[ASTextNode placeholderImage]+357>
	     <-[ASDisplayNode _completeLayoutCalculation]+558>
	      <-[ASDisplayNode measureWithSizeRange:]+794>
	       <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	        <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objec
	         <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChildrenA
	          <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obje
	           <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0_9allo
	            <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	             <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc
	               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChi
	                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11obj
	                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0
	                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                   <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                    <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                     <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_ob   <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc
	               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChi
	                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11obj
	                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0
	                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                   <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                    <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                     <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46lay
	                      <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutabl
	                       <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obj
	                        <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                         <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                          <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                           <-[ASDisplayNode measureWithSizeRange:]+321>
	                            <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                             <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12AS
	                               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11Jul 26 15:18:02  BroEx[6115] <Error>: CGContextSetFillColorWithColor: invalid context 0x0. Backtrace:
	  <CGContextSetFillColorWithColor+66>
	   <-[UIDeviceWhiteColor setFill]+55>
	    <-[ASTextNode placeholderImage]+357>
	     <-[ASDisplayNode _completeLayoutCalculation]+558>
	      <-[ASDisplayNode measureWithSizeRange:]+794>
	       <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	        <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objec
	         <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChildrenA
	          <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obje
	           <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0_9allo
	            <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	             <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc
	               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46layoutChi
	                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11obj
	                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS0
	                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                   <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                    <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                     <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_EEEEZL46lay
	                      <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutabl
	                       <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obj
	                        <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                         <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                          <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                           <-[ASDisplayNode measureWithSizeRange:]+321>
	                            <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                             <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12AS
	                               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_E
	                                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12A
	                                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                                   <-[ASBackgroundLayoutSpec measureWithSizeRange:]+152>
	                                    <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                                     <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                                      <-[ASDisplayNode measureWithSizeRange:]+321>
	                                       <-[ASDataController _layoutNode:withConstrainedSize:]+161>
	                                        <-[ASDataController _layoutNodes:fromContexts:atIndexesOfRange:ofKind:]+1287>
	                                         <__63-[ASDataController _layoutNodesFromContexts:ofKind:completion:]_block_invoke.121+
	                                          <_dispatch_call_block_and_release+15>
	                                           <_dispatch_client_callout+14>
	                                            <_dispatch_root_queue_drain+1008>
	                                             <_dispatch_worker_thread3+115>
	                                              <_pthread_wqthread+1050>
    objc_objectNS1_9allocatorIS5_E
	                                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12A
	                                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                                   <-[ASBackgroundLayoutSpec measureWithSizeRange:]+152>
	                                    <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                                     <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                                      <-[ASDisplayNode measureWithSizeRange:]+321>
	                                       <-[ASDataController _layoutNode:withConstrainedSize:]+161>
	                                        <-[ASDataController _layoutNodes:fromContexts:atIndexesOfRange:ofKind:]+1287>
	                                         <__63-[ASDataController _layoutNodesFromContexts:ofKind:completion:]_block_invoke.121+
	                                          <_dispatch_call_block_and_release+15>
	                                           <_dispatch_client_callout+14>
	                                            <_dispatch_root_queue_drain+1008>
	                                             <_dispatch_worker_thread3+115>
	                                              <_pthread_wqthread+1050>
     jectNS1_9allocatorIS5_EEEEZL46lay
	                      <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutabl
	                       <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_obj
	                        <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                         <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                          <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                           <-[ASDisplayNode measureWithSizeRange:]+321>
	                            <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                             <_ZL16crossChildLayoutPU23objcproto12ASLayoutable11objc_object22ASStackLayoutSpecStyleffff+372>
	                              <_ZZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12AS
	                               <_ZN2AS3mapINSt3__16vectorIU8__strongPU23objcproto12ASLayoutable11objc_objectNS1_9allocatorIS5_E
	                                <_ZL46layoutChildrenAlongUnconstrainedStackDimensionRKNSt3__16vectorIU8__strongPU23objcproto12A
	                                 <_ZN25ASStackUnpositionedLayout7computeERKNSt3__16vectorIU8__strongPU23objcproto12ASLayoutable
	                                  <-[ASStackLayoutSpec measureWithSizeRange:]+1644>
	                                   <-[ASBackgroundLayoutSpec measureWithSizeRange:]+152>
	                                    <-[ASInsetLayoutSpec measureWithSizeRange:]+1741>
	                                     <-[ASDisplayNode calculateLayoutThatFits:]+1262>
	                                      <-[ASDisplayNode measureWithSizeRange:]+321>
	                                       <-[ASDataController _layoutNode:withConstrainedSize:]+161>
	                                        <-[ASDataController _layoutNodes:fromContexts:atIndexesOfRange:ofKind:]+1287>
	                                         <__63-[ASDataController _layoutNodesFromContexts:ofKind:completion:]_block_invoke.121+
	                                          <_dispatch_call_block_and_release+15>
	                                           <_dispatch_client_callout+14>
	                                            <_dispatch_root_queue_drain+1008>
	                                             <_dispatch_worker_thread3+115>
	                                              <_pthread_wqthread+1050>

### PR DESCRIPTION
**Pull request by hannahmbanana**
This fixes #1986.

screen shot 2016-08-14 at 12 08 13 pm

### CODE REVIEW

File: AsyncDisplayKit/ASTextNode.mm [hunk 1]
```diff
@@ -903,7 +903,7 @@ - (UIImage *)placeholderImage
   // FIXME: Replace this implementation with reusable CALayers that have .backgroundColor set.
   // This would completely eliminate the memory and performance cost of the backing store.
   CGSize size = self.calculatedSize;
-  if (CGSizeEqualToSize(size, CGSizeZero)) {
+  if (size.width * size.height < FLT_EPSILON) {
```
Comment (maicki) [inline AsyncDisplayKit/ASTextNode.mm]: For better readiness could we add `()` around `size.width * size.height`?

**Comment by hannahmbanana** [comment]
@Adlai-Holler: an easy one

**Comment by Adlai-Holler** [comment]
This is awesome, but let's be super precise. FLT_EPSILON is for single-precision floats, but on 64-bit CGFloats are actually double-precision.

Add the define from this gist https://gist.github.com/Kentzo/1985919 into CGRect+ASConvenience.h, and use CGFLOAT_EPSILON instead.

**Comment by Adlai-Holler** [comment]
🌴 ☀️

**Comment by Adlai-Holler** [comment]
@hannahmbanana Let's make sure this gets into our 1.x compat releases so nobody is left out.

**Pull request merged**